### PR TITLE
cinn(backends): generate infer shape kernel to infer shape of output tensor

### PR DESCRIPTION
**Pull request by 6clc**
### PR types

Others

### PR changes

Others

### Description

Pcard-78120

通过二维指针来返回后端infer shape的结果。生成的cinn ir如下。tensor_shape_args是一个二维指针。  infer_shape_set_value(0, 0, S1, tensor_shape_args) 表示将第0个output tensor的第0维设置为S1。
```cpp
|| 633: function fn_exp_subtract_infer_shape (kernel_args, kernel_args_num, tensor_shape_args)
|| 633: {
|| 633:   int32 S0 = cinn_get_value_in_cuda_kernel_args(kernel_args, 3)
|| 633:   int32 S1 = cinn_get_value_in_cuda_kernel_args(kernel_args, 4)
|| 633:   {
|| 633:     infer_shape_set_value(0, 0, S1, tensor_shape_args)
|| 633:     infer_shape_set_value(0, 1, 128, tensor_shape_args)
|| 633:     infer_shape_set_value(1, 0, S1, tensor_shape_args)
|| 633:     infer_shape_set_value(1, 1, 128, tensor_shape_args)
|| 633:   }
|| 633: }

```

### CODE REVIEW

File: paddle/cinn/backends/codegen_cuda_host.cc [hunk 1]
```diff
@@ -198,13 +198,23 @@ llvm::Value* CodeGenCUDA_Host::LowerHostFunc(const ir::_LoweredFunc_* func) {
                  [](auto& arg) { return std::addressof(arg); });
   // @}
 
+  // Set local scope table
+  CHECK_EQ(ll_function_args.size(), func->args.size());
+  for (int i = 0; i < ll_function_args.size(); i++) {
```
Comment (zhhsplendid) [inline paddle/cinn/backends/codegen_cuda_host.cc]: Prefer ++i 
Reference: https://google.github.io/styleguide/cppguide.html#Preincrement_and_Predecrement

File: paddle/cinn/hlir/framework/pir/op_lowering_impl.cc [hunk 2]
```diff
@@ -1023,6 +1034,47 @@ bool OpLowererImpl::IsInTensorMap(
...
 
+ir::LoweredFunc OpLowererImpl::GenerateInferShapeFunc(
+    const GroupPtr& group,
+    const std::vector<ir::Tensor> group_func_arg_tensors,
+    const std::vector<ir::Argument> group_func_args) {
+  // CHECK_EQ(group_func_arg_tensors.size(), group_func_args.size());
+  std::vector<ir::Expr> ir_bodys;
+  int output_tensor_idx = 0;
+  for (int tensor_arg_idx = 0; tensor_arg_idx < group_func_arg_tensors.size();
+       tensor_arg_idx++) {
```
Comment (zhhsplendid) [inline paddle/cinn/hlir/framework/pir/op_lowering_impl.cc]: Same as above

File: paddle/fluid/framework/new_executor/instruction/cinn_jit_instruction.cc [hunk 3]
```diff
@@ -61,6 +62,49 @@ class CinnJitInstruction::FnPtrImpl {
...
+      func_args_.emplace_back(kernel_args[int_arg_mp.second.arg_idx]->dims().at(
+          int_arg_mp.second.dim_idx));
+      func_args_.emplace_back(static_cast<int64_t>(
+          kernel_args[int_arg_mp.second.arg_idx]->dims().at(
+              int_arg_mp.second.dim_idx)));
+    }
+
+    // 3. Define an array of Pointers to hold the output tensor shape
+    int32_t* output_tensor_shapes[output_tensor_size];
+    for (int i = 0; i < output_tensor_size; i++) {
```
Comment (zhhsplendid) [inline paddle/fluid/framework/new_executor/instruction/cinn_jit_instruction.cc]: Prefer ++i

File: paddle/fluid/framework/new_executor/instruction/cinn_jit_instruction.cc [hunk 4]
```diff
@@ -61,6 +62,49 @@ class CinnJitInstruction::FnPtrImpl {
...
+    }
+
+    // 4. Launch infer_shape_fn_ptr to infer shape of output tensor
+    ((infer_shape_func_ptr_g)cinn_kernel_info_.infer_shape_fn_ptr)(
+        static_cast<void*>(func_args_.data()),
+        func_args_.size(),
+        output_tensor_shapes);
+
+    // 5. Resize shape of output tensor
+    for (int i = 0; i < output_tensor_size; i++) {
```
Comment (zhhsplendid) [inline paddle/fluid/framework/new_executor/instruction/cinn_jit_instruction.cc]: Prefer ++i

**Review by zhangbo9674 — APPROVED** [review]
LGTM for jit_instruction

File: paddle/cinn/backends/codegen_cuda_host.cc (hunk 1 above)
Comment (6clc) [reply paddle/cinn/backends/codegen_cuda_host.cc]: done

File: paddle/cinn/hlir/framework/pir/op_lowering_impl.cc (hunk 2 above)
Comment (6clc) [reply paddle/cinn/hlir/framework/pir/op_lowering_impl.cc]: done

File: paddle/fluid/framework/new_executor/instruction/cinn_jit_instruction.cc (hunk 3 above)
Comment (6clc) [reply paddle/fluid/framework/new_executor/instruction/cinn_jit_instruction.cc]: done

File: paddle/fluid/framework/new_executor/instruction/cinn_jit_instruction.cc (hunk 4 above)
Comment (6clc) [reply paddle/fluid/framework/new_executor/instruction/cinn_jit_instruction.cc]: done

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Pull request merged**